x/exchange/market.go

```diff
@@ -0,0 +1,662 @@
+package exchange
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+
+	sdkmath "cosmossdk.io/math"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+
+	nametypes "github.com/provenance-io/provenance/x/name/types"
+)
+
+const (
+	// MaxName is the maximum length of MarketDetails.Name
+	MaxName = 250
+	// MaxDescription is the maximum length of MarketDetails.Description
+	MaxDescription = 2000
+	// MaxWebsiteURL is the maximum length of MarketDetails.WebsiteUrl
+	MaxWebsiteURL = 200
+	// MaxIconURI is the maximum length of MarketDetails.IconUri
+	MaxIconURI = 2000
+)
+
+var (
+	_ authtypes.AccountI       = (*MarketAccount)(nil)
+	_ authtypes.GenesisAccount = (*MarketAccount)(nil)
+)
+
+// Validate makes sure that everything in this Market is valid.
+// The MarketId is allowed to be zero in here.
+// Some uses might require it to have a value, but that check is left up to the caller.
+func (m Market) Validate() error {
+	return errors.Join(
+		// Nothing to check on the MarketId. It's allowed to be zero to indicate to use the next one.
+		m.MarketDetails.Validate(),
+		ValidateFeeOptions("create-ask flat fee", m.FeeCreateAskFlat),
+		ValidateFeeOptions("create-bid flat fee", m.FeeCreateBidFlat),
+		ValidateFeeOptions("seller settlement flat fee", m.FeeSellerSettlementFlat),
+		ValidateFeeOptions("buyer settlement flat fee", m.FeeBuyerSettlementFlat),
+		ValidateFeeRatios(m.FeeSellerSettlementRatios, m.FeeBuyerSettlementRatios),
+		ValidateAccessGrantsField("", m.AccessGrants),
+		// Nothing to check for with the AcceptingOrders and AllowUserSettlement booleans.
+		ValidateReqAttrs("create-ask", m.ReqAttrCreateAsk),
+		ValidateReqAttrs("create-bid", m.ReqAttrCreateBid),
+	)
+}
+
+// ValidateFeeOptions returns an error if any of the provide coin values is not a valid fee option.
+func ValidateFeeOptions(field string, options []sdk.Coin) error {
+	var errs []error
+	denoms := make(map[string]bool)
```

Comment:
This is very minor, and really up to you if you want to include it. You can specify a minimum size for the map so that it potentially won't have to reallocate.

denoms := make(map[string]bool, len(options))
dups := make(map[string]bool, len(options))

Reply:
I've never really looked into the guts of how maps grow and when it's beneficial to include a length with a `make(map...)`. I've known it's possible, but haven't ever seen any discussions about when its best to include it. By contrast, I've seen lots of discussions about including the length and optional capacity when creating slices. I guess I've just assumed the make function needed to allow the lengths for slices, and therefore also allows a length for maps even if it doesn't really do anything (even though I don't actually know it doesn't really do anything).

Reply:
Yeah, the map hashes items into an array and each of the array elements are a list of buckets. As those buckets are filled the map automatically double sthe number of buckets and redistributes the keys. I doubt we will see a noticeable difference (depending on how big options can get), but pointed it out since I saw it.

Reply:
Heh. From https://go.dev/ref/spec#Making_slices_maps_and_channels

> Calling make with a map type and size hint n will create a map with initial space to hold n map elements. The precise behavior is implementation-dependent.

They say in a few places that `make(T, n)` where `T` is a `map` means "map of type T with initial space for approximately n elements"

Reply:
All that being said, I did provide suggested sizes where it made sense to when making maps in all this exchange stuff.